p2p/net/swarm/dial_worker.go

```diff
@@ -159,6 +170,8 @@ loop:
 		//     interested in dials on this address.
 
 		select {
+		case <-w.ctx.Done():
+			return
```

Comment:
This doesn't have the metrics tracing code. Can you add that bit? 
```
		case req, ok := <-w.reqch:
			if !ok {
				if w.s.metricsTracer != nil {
					w.s.metricsTracer.DialCompleted(w.connected, totalDials)
				}
				return
			}
```

Reply:
Handled by the rebase.